src/front/glsl/builtins.rs

```diff
@@ -1429,88 +1400,113 @@ impl MacroCall {
                 ctx.samplers.insert(args[0], args[1]);
                 Ok(args[0])
             }
-            MacroCall::Texture | MacroCall::TextureLod => {
-                let comps = parser.coordinate_components(ctx, args[0], args[1], meta, body)?;
-                let level = match *self {
-                    MacroCall::Texture => args
-                        .get(2)
-                        .copied()
-                        .map_or(SampleLevel::Auto, SampleLevel::Bias),
-                    MacroCall::TextureLod => SampleLevel::Exact(args[2]),
-                    _ => unreachable!(),
-                };
-                texture_call(ctx, args[0], level, comps, body, meta)
-            }
-            MacroCall::TextureGrad(shadow) => {
-                let comps = parser.coordinate_components(ctx, args[0], args[1], meta, body)?;
-                let level = match shadow {
+            MacroCall::Texture{ proj, offset, level_type } => {
+                let coords = match proj {
```

Comment:
nit: this could be rewritten as
```rust
let mut coord = args[1];
if proj {
...
}
```
I think it symbolizes better that argument 1 is the coordinate vector and reduces one level of indentation